extra/plugins/dioptra_optic/data.py

```diff
@@ -0,0 +1,218 @@
+        See `_load_dataset_from_builder` for description of arguments
+
+    Returns:
+        A tuple of train, val, test tf.data.Dataset objects.
+    """
+    builder = tfds.builder(name, data_dir=data_dir)
+    return _load_dataset_from_builder(
+        builder,
+        split=split,
+        normalize_val=normalize_val,
+        image_size=image_size,
+        batch_size=batch_size,
+        seed=seed,
+    )
+
+
+@pyplugs.register
+def load_dataset_from_image_directory(
+    data_dir: Path,
+    split: dict[Literal["train", "val", "test"], str] = {
+        "train": "train[:80%]",
+        "val": "train[80%:]",
+        "test": "test",
+    },
+    normalize_val: float | None = 255.0,
+    image_size: tuple[int, int] | None = None,
+    batch_size: int = 32,
+    seed: int | None = None,
+) -> tuple[tf.data.Dataset | None, tf.data.Dataset | None, tf.data.Dataset | None]:
+    """
+    Loads a tfds dataset by path from disk and constructs a preprocessing pipeline.
+
+    Args:
+        data_dir: The path to the directory containing the image dataset
+        See `_load_dataset_from_builder` for description of arguments
+
+    Returns:
+        A tuple of train, val, test tf.data.Dataset objects.
+    """
+    builder = tfds.ImageFolder(str(data_dir))
+    return _load_dataset_from_builder(
+        builder,
+        split=split,
+        normalize_val=normalize_val,
+        image_size=image_size,
+        batch_size=batch_size,
+        seed=seed,
+    )
+
+
+def _load_dataset_from_builder(
+    builder: tfds.core.DatasetBuilder,
+    split: dict[Literal["train", "val", "test"], str] = {
+        "train": "train[:80%]",
+        "val": "train[80%:]",
+        "test": "test",
+    },
+    normalize_val: float | None = 255.0,
+    image_size: tuple[int, int] | None = None,
+    batch_size: int = 32,
+    seed: int | None = None,
+) -> tuple[tf.data.Dataset | None, tf.data.Dataset | None, tf.data.Dataset | None]:
+    """
+    Args:
+        builder: A tfds.core.DatasetBuilder to build the dataset from.
+        split: A tuple denoting the train/val/test split for the dataset.
+            See https://www.tensorflow.org/datasets/splits
+        normalize_val: The data is normalized by dividing by this value.
+            If None, no data normalization is applied
+        image_size: Size to
+            If None, no resizing is applied.
+        batch_size: Applies batching to the dataset with the specified size.
+            If None, no batching is applied.
+        seed: The random seed used to shuffle the train split. Used for deterministic results.
+            If None, shuff
+
+    Returns:
+        A tuple of train, val, test tf.data.Dataset objects.
+    """
+    # get a dictionary of dataset splits from the builder
+    dataset: dict = builder.as_dataset(split=split, as_supervised=True)
+
+    def normalize(x, y):
+        """Normalizes images: `uint8` -> `float32`."""
+        return tf.cast(x, tf.float32) / normalize_val, y
```

Comment:
Explicitly cast all non-Tensor values to desired datatype before running division op:

```suggestion
        return tf.cast(x, tf.float32) / tf.cast(normalize_val, tf.float32), y
```